### ISSUE TITLE
RAM issues: we need tokenization

### ISSUE THREAD

**Issue by hmedina** [Kappa-Dev/KaSim#228]
Models with 1:1 abundances of biological components require high RAM capacity. My Wnt model, at 20:1, uses over 17GB, so running the full model would require at least 340GB.

As we have seen from assemblies, monomers tend to be the most abundant species, so tokenization could reduce the RAM footprint tremendously.

**Comment by jkrivine** [comment]
Hi Hector
Do you mean automated tokenification?
If not then you can use actual tokens and implement transition rules 
token_monomers <-> graph dimer

**Comment by hmedina** [comment]
Apologies, I mean automatic tokenification.

The model contains agents with many sites, APC alone has over 50 sites to
phosphorylate and over 10 binding sites, so manually making the
10*2^50=1.1258999e+16 binding/unbinding rules for all APC monomers would be
not desirable... I do however have a script for programmatically expanding
site combinatorics, but a text file with 1e16 lines seems a bit
excessive... I'll try on the server instead of my laptop >:)

I imagine automatic tokenizarion would be done lazily, i.e. only when the
monomer is present in the mixture does the token get generated. Otherwise,
if created at initialization, it would create many tokens that might go
unused, e.g. kinetically irrelevant, though I do not know if that would
have a perceivable impact. Is 1e16 too many tokens?

**Comment by feret** [comment]
Dear Hector, 

That is amazing. 
I had in min a lazy version of tokens, where token would be created at demand, but not garbage collected. 

However in your case, it could be worth looking if the monomers could be encoded more compactly. 
As bitfield as a first try, or even using Huffman compression trick. 

Cheers.
Jérôme.

**Comment by feret** [comment]
Dear Hector, 

It seems somehow that ACP is made of 7 independent blocks of sites:
Agent_trace_APC_S1272~_S1275~_S1276~_S1278~_S1281~_E1284~_rpt20AA1!.dot                           Agent_trace_APC_S1386~_T1389~_S1390~_S1392~_D1395~_E1398~_rpt20AA2!.dot                           Agent_trace_APC_S1501~_S1504~_S1505~_S1507~_S1510~_E1513~_rpt20AA3!.dot                           Agent_trace_APC_S1652~_T1655~_S1656~_S1658~_T1661~_S1664~_rpt20AA4!.dot                           Agent_trace_APC_S1857~_D1860~_S1861~_S1863~_D1866~_D1969~_rpt20AA5!.dot                           Agent_trace_APC_S1964~_S1967~_S1968~_S1970~_S1973~_D1976~_rpt20AA6!.dot                           Agent_trace_APC_S2022~_S2025~_S2026~_S2028~_S2031~_S2034~_rpt20AA7!.dot

This seems to suggest that CTMC-fragments could help to get a rationale tokenization. 
We can track the population of fragments, instead of the population chemical species without disturbing the CTMC. 
Then we would be left we 7*2^7 kind of tokens, instead of 2^50. 

Usually CTMC-fragments does not work, because the soundness requirements are too strong. 
Do you think it may be due to some missing interactions in your model ? 
Or maybe this is hidden in the fact, that connectiveness check (i.e. unary VS binary rule) are not yet supported in the analysis of the flow of information, and if you consider it, fragments would be merge in one single macro-protein.

Cheers.
Jérôme.

**Comment by hmedina** [comment]
Hi Jerome,
I don't understand why the CTMC fragments do not work; more precisely, what it means for them to not work. Are you saying the set of CTMC fragments is not self-consistent? Or that its union does not map to the contact map?

I'm curious as to what this failure means, because perhaps model parts could be re-written in a different style (e.g. to have more interactions).

Best,
Héctor

**Comment by feret** [comment]
Let us take an example. 

A(x),B(x) -> A(x!1),B(x!1)
A(y),C(y) -> A(y!1),C(y!1)

The CTMC fragmentation of the model will give this one:

Ax(x),B(x) -> Ax(x!1),B(x!1)
Ay(y),C(y) -> Ay(y!1),C(y!1)

ensuring that if you take an Ax randomly and an Ay randomly, the probability that they belong to the same A is 1 over the number of As. 
This property is crucial, this is the induction hypothesis that is used to prove that the CTMC that you obtain in the latter model is the exact projection of the one you obtain in the former one.

Now add the rule:
B(),C() -> B() @1,2

You can fairly compute the event distribution at a given abstract state provided that the induction hypothesis is satisfied. 
But is the induction hypothesis still satisfied ? 
I would say no. 

In other words, you have to update the definition of the syntactic abstraction of flow of information to cover the flow coming from connectedness. 
For instance, by saying that if a rule can test the connectedness between two agents, fragmentation should preserve this connectedness. 
But then, obviously you would fragment less. 

Does it help? 

Cheers
Jérôme.

**Comment by pirbo** [comment]
What to do with this issue?
Hector issue had nothing to do with the need of tokenification but with the silliness of the old algorithm to deal with unimolecular rules. This piece of code has been replaced so this exact situation has gone.
General discussion that followed (even if it refers to a non public non published model :-)) stays interesting and is not closed...

**Comment by feret** [comment]
Yes, it must be the 10 years anniversary of the first time we said tokenization is important. 
It is not so interesting from a research point of view with respect to what we are working on, and we lack of man power. 

Yet it is definitely interesting and is quite likely to open the door to many issues related to hybridization and conservative abstractions. 

Also, I am not so sure the issue list in github is the proper place to write about it.

**Comment by pirbo** [comment]
Let's close here then
